In [123]:
# Задание 1. Загрузка данных¶
# Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 
# столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, 
# pain, outcome.

In [124]:
import pandas as pd
df = pd.read_csv('horse_data.csv', header=None, na_values = '?')
df8 = df[[(0), (1), (3), (4), (5), (6), (10), (22)]]
df8.columns = ['Surgery', 'Age', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 'outcome']
df8

Surgery  Age  rectal temperature  pulse  respiratory rate  \
0        2.0    1                38.5   66.0              28.0   
1        1.0    1                39.2   88.0              20.0   
2        2.0    1                38.3   40.0              24.0   
3        1.0    9                39.1  164.0              84.0   
4        2.0    1                37.3  104.0              35.0   
..       ...  ...                 ...    ...               ...   
295      1.0    1                 NaN  120.0              70.0   
296      2.0    1                37.2   72.0              24.0   
297      1.0    1                37.5   72.0              30.0   
298      1.0    1                36.5  100.0              24.0   
299      1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           NaN   NaN      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[300 rows x 8 columns]

##### 1:  surgery
          1 = Yes, it had surgery
          2 = It was treated without surgery
##### 2:  Age
          1 = Adult horse
          2 = Young (< 6 months)
##### 3:  rectal temperature
          - linear
          - in degrees celsius.
          - An elevated temp may occur due to infection.
          - temperature may be reduced when the animal is in late shock
          - normal temp is 37.8
          - this parameter will usually change as the problem progresses
               eg. may start out normal, then become elevated because of
                   the lesion, passing back through the normal range as the
                   horse goes into shock
##### 4:  pulse
          - linear
          - the heart rate in beats per minute
          - is a reflection of the heart condition: 30 -40 is normal for adults
          - rare to have a lower than normal rate although athletic horses
            may have a rate of 20-25
          - animals with painful lesions or suffering from circulatory shock
            may have an elevated heart rate
##### 5:  respiratory rate
          - linear
          - normal rate is 8 to 10
          - usefulness is doubtful due to the great fluctuations
##### 6:  temperature of extremities
          - a subjective indication of peripheral circulation
          - possible values:
               1 = Normal
               2 = Warm
               3 = Cool
               4 = Cold
          - cool to cold extremities indicate possible shock
          - hot extremities should correlate with an elevated rectal temp.
##### 7: pain 
          - a subjective judgement of the horse's pain level
          - possible values:
               1 = alert, no pain
               2 = depressed
               3 = intermittent mild pain
               4 = intermittent severe pain
               5 = continuous severe pain
          - should NOT be treated as a ordered or discrete variable!
          - In general, the more painful, the more likely it is to require
            surgery
          - prior treatment of pain may mask the pain level to some extent
##### 8: outcome
          - what eventually happened to the horse?
          - possible values:
               1 = lived
               2 = died
               3 = was euthanized

In [125]:
# Задание 2. Первичное изучение данных
# Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

 Для начало соберем общую статистику по всем столбцам, чтобы понять какие из столбцов заполнены верно,а к каким есть вопросы. Для столбцов с количественными характеристиками(rectal temperature, pulse, respiratory rate)  используем метод .describe(), а для столбцов с качественными (surgery, Age, temperature of extremities,pain, outcome) .value_counts

In [126]:
df8.describe()

Surgery         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

In [127]:
print(df8['Surgery'].value_counts(), '\n')
print(df8['Age'].value_counts(), '\n')
print(df8['temperature of extremities'].value_counts(), '\n')
print(df8['pain'].value_counts(), '\n')
print(df8['outcome'].value_counts(), '\n')

1.0    180
2.0    119
Name: Surgery, dtype: int64 

1    276
9     24
Name: Age, dtype: int64 

3.0    109
1.0     78
2.0     30
4.0     27
Name: temperature of extremities, dtype: int64 

3.0    67
2.0    59
5.0    42
4.0    39
1.0    38
Name: pain, dtype: int64 

1.0    178
2.0     77
3.0     44
Name: outcome, dtype: int64 



Полученные данные сопоставляем с описанием столюцов и я вижу следующие проблемы в датасете

 1) Столбец Age предполагает наличие в нем данных - 1 или 2(взрослый или нет), у нас имеет место быть значение 9 что как мне кажется является опечаткой и вместо 9 должна стоять 2, если все лошади были бы у нас взрослыми добавлять такой столбец в дата сет было бы не целесообразно, а значит статистика молодых особей тоже должна быть
 
2) Столбец pulse в описании дата сета стоит что нормальный лошадиный пульc варируется в зависитмоти от возраста и физического состаяния лошади в диапазоне от 20 до 40 уд/мин, так как у нас идут данные по лошадям испытывающим проблемы со здоровьем логически можно предположить что пульс может быть выше и медиана в наших данных составляет 64 уд/мин, но  в дата сете у нас максимум в этом столбце составляет 184 уд/мин, что как мне кажется является очевидным выбросом, этот стобец нужно проверить на выбросы.

3) Столбец respiratory rate по аналогии со столбцом pulse, в описание дата сета фигурирует что норма 8-10, медиана 24.5, максимум 96, но в описании датасета указано что этот показатель имеет большой разброс, так что возможно здесь выбросов нет.

По остальным столбцам все данные находятся в пределах и согласно описанию датасета

In [128]:
# Задание 3. Работа с пропусками
# Рассчитать количество пропусков для всех выбранных столбцов. 
# Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных 
# статистик и возможной взаимосвязи значений в них. 
# Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [129]:
df8.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Surgery                     299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory rate            242 non-null    float64
 5   temperature of extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [130]:
df8[df8['Surgery'].isnull()] 

Surgery  Age  rectal temperature  pulse  respiratory rate  \
132      NaN    1                38.0   48.0              20.0   

     temperature of extremities  pain  outcome  
132                         3.0   4.0      NaN

Age  - пропусков нет
 
 Surgery , outcome - пропуск один в одной и тойже строчке, судя по остальным показателем скорее всего данной особи могла бы
быть проведена операция, но данных по результату операции тоже нет, и каких либо выводов по успешности операции мы получить не
можем так что эту строчку я бы просто удалил

rectal temperature - по данным полученным в прошлом задании можно сказать в общем все температура имеет минимальный разброс так что данный пункт можно заполнить медианным значением
    
temperature of extremities - в описании данных сказано что данная температура имеет взаимосвязь с rectal temperature, используем медианное значение на основании rectal temperature 

respiratory rate - в описании сказано что что это сомнительные показатели, 
так как имеют слишком большой разброс значений, их так же можно заполнить медианным показателем 
    
pain - как указано в описании высокий показатель боли напрямую ведет к операции заполняем медианным значением на основании Surgery        
 
pulse - опять же по описанию зависит от наличия боли заполняем медианным показателем на основании pain

In [131]:
ndf = df8.drop(index=[132])
ndf['rectal temperature'].fillna(ndf['rectal temperature'].median(), inplace=True)
ndf['temperature of extremities'].fillna(ndf.groupby(['rectal temperature'])['temperature of extremities'].transform('median'), inplace=True)
ndf['respiratory rate'].fillna(ndf['respiratory rate'].median(), inplace=True)
ndf['pain'].fillna(ndf.groupby(['Surgery'])['pain'].transform('median'), inplace=True)
ndf['pulse'].fillna(ndf.groupby(['pain'])['pulse'].transform('median'), inplace=True)

In [132]:
ndf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Surgery                     299 non-null    float64
 1   Age                         299 non-null    int64  
 2   rectal temperature          299 non-null    float64
 3   pulse                       299 non-null    float64
 4   respiratory rate            299 non-null    float64
 5   temperature of extremities  297 non-null    float64
 6   pain                        299 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 21.0 KB


In [120]:
# Две строчки почему-то не заполнил
ndf[ndf['temperature of extremities'].isnull()] 

Surgery  Age  rectal temperature  pulse  respiratory rate  \
113      1.0    1                36.8   60.0              28.0   
141      2.0    1                36.0   42.0              30.0   

     temperature of extremities  pain  outcome  
113                         NaN   3.0      2.0  
141                         NaN   2.0      2.0

In [136]:
ndf['temperature of extremities'].describe()

count    297.000000
mean       2.375421
std        1.004871
min        1.000000
25%        1.000000
50%        3.000000
75%        3.000000
max        4.000000
Name: temperature of extremities, dtype: float64